core/src/runtime/record.rs

```diff
@@ -434,15 +438,29 @@ impl MachineRecord for ExecutionRecord {
             .program_memory_events
             .extend_from_slice(&self.program_memory_events);
 
+        // Set the pi_digest for all shards.
+        for shard in shards.iter_mut() {
+            shard.set_pi_digest(self.pi_digest.unwrap());
+        }
+
         shards
     }
+
+    fn pi_digest(&self) -> Option<PiDigest<u32>> {
```

Comment:
`pi_digest` and `set_pi_digest` was added to the `MachineRecord` trait so that both the `core` and `recursion`'s `ExecutionRecord` will implement it.

This actually was a suggestions from @jtguibas